# PhyliCS Usage
This repository contains the data and the jupyter notebooks to reproduce the case studies presented in [PhyliCS](https://github.com/bioinformatics-polito/PhyliCS/tree/master) paper.

Spefifically, 
- [synth.ipynb](https://github.com/bioinformatics-polito/PhyliCS_usage/blob/main/synth.ipynb) implements the code to compute SHscore statistics presented in the section **Experiment 1: SHscore on synthetic data**
- [shscores_multiple_sizes.ipynb](https://github.com/bioinformatics-polito/PhyliCS_usage/blob/main/shscores_multiple_sizes.ipynb) implements the code to reproduce the results shown in the section **Experiment 1: SHscore on synthetic data**, subsection **SHscore indipendence from CNA size and gained copy number** of the paper
- [shscore_mrca.ipynb](https://github.com/bioinformatics-polito/PhyliCS_usage/blob/main/shscore_mrca.ipynb) implements the code to reproduce the results shown in the section **Experiment 2: SHscore and evolutionary distance** 
- [breast_tumor.ipynb](https://github.com/bioinformatics-polito/PhyliCS_usage/blob/main/breast_tumor.ipynb) implements the code to reproduce the analysis on 10x Genomics breast tumor data presented in section **Experiment 3: SHscore on tumor data**, subsection **Spatial subsamples from the same disease site**
- [lung_tumor.ipynb](https://github.com/bioinformatics-polito/PhyliCS_usage/blob/main/lung_tumor.ipynb) implements the code to reproduce the analysis on public lung primary tumors and liver metastasis data presented in section **Experiment 3: SHscore on tumor data**, subsection **Spatial subsamples from the different disease sites**
- [mdamb231.md](https://github.com/bioinformatics-polito/PhyliCS_usage/blob/main/mdamb231.md) contains the code to reproduce the analysis on the MDA-MB-231 cell line and on the two daughter cell lines, MDA-MB-231-EX1 and MDA-MB-231-EX2, obtained from the clonal expansion of two cells. Here we present only PhyliCS application, while you can find the pipeline to download and allign the sequencing reads and to call the scCNA events at [Snakefile](https://github.com/bioinformatics-polito/PhyliCS_usage/tree/main/snakemake/Snakefile_mdamb231_cellline). For this reason, for this particular case we provide a markdown static file, rather that a jupyter notebook
- [data](https://github.com/bioinformatics-polito/PhyliCS_usage/tree/main/data) contains all data needed to reproduce the analyses, with the exception on MDA-MB-231 data which can't be deposited at this repository due to IRB protocol. Raw datasets are deposited under BioProject PRJNA629885
- [snakemake](https://github.com/bioinformatics-polito/PhyliCS_usage/tree/main/snakemake) contains all data needed to reproduce the alignment and the CNA calling procedures for MDA-MB-231 data
- [run_ginkgo](https://github.com/bioinformatics-polito/PhyliCS_usage/tree/main/run_ginkgo) contains the code of an old version of PhyliCS, which contains the implementation of a standalone version of Ginkgo, which we used to perform scCNA calling on MDA-MB-231 data

In order to read the notebooks content, you just need to click on them and they will automatically be rendered as HTML files. If you prefer, you may clone the current directory and execute the notebooks locally, after [installing PhyliCS](https://github.com/bioinformatics-polito/PhyliCS#installation-and-setup-instructions) and its dependencies.

If you encounter any problem in rendering the notebooks you may navigate to each `data` subfolder where you can find a README describing the use case (e.g. `data\simulations\README.md`).
